app/src/main/java/de/dhbw/corona_world_app/api/APIManager.java

```diff
@@ -126,28 +130,50 @@ public Country call() throws Exception {
             Logger.logV(TAG, "Country-List finished constructing...");
         } else {
             Logger.logE(TAG, "Throwing IllegalArgumentException! MAX_COUNTRY_LIST_SIZE has been exceeded!");
-            throw new IllegalArgumentException("Input country list is too big, max allowed " + MAX_COUNTRY_LIST_SIZE +"!");
+            throw new IllegalArgumentException("Input country list is too big, max allowed " + MAX_COUNTRY_LIST_SIZE + "!");
         }
         return returnList;
     }
 
-    public static List<TimeframedCountry> getData(List<ISOCountry> countryList, List<Criteria> criteriaList, LocalDate startDate, LocalDate endDate) throws ExecutionException, InterruptedException, JSONException {
+    public static List<TimeframedCountry> getData(@NonNull List<ISOCountry> countryList, @NonNull List<Criteria> criteriaList, LocalDate startDate, LocalDate endDate) throws ExecutionException, InterruptedException, JSONException {
         Logger.logV(TAG, "Getting data according to following parameters: " + countryList + " ; " + criteriaList);
+        if ((startDate == null && endDate != null))
+            throw new IllegalArgumentException("Ending date is before starting date!");
+        if (startDate != null && endDate != null && endDate.isBefore(startDate))
+            throw new IllegalArgumentException("Ending date is before starting date!");
         List<TimeframedCountry> returnList = new ArrayList<>();
         List<Future<String>> futureCoronaData = new ArrayList<>();
         List<Future<Country>> futurePopData = new ArrayList<>();
+        if (startDate == null) startDate = LocalDate.now();
+        if (endDate == null) endDate = LocalDate.now();
+
+        final LocalDate finalStartDate;
+        final LocalDate finalEndDate = endDate;
+
+        //this is only needed because the api cannot handle when start and end date are equal and returns all dates' data
+        boolean startAndEndEqual = startDate.equals(endDate);
+        if(startAndEndEqual){
+            finalStartDate = startDate.minusDays(1);
+        } else {
+            finalStartDate = startDate;
+        }
```

Comment:
Finalstart- and enddate seem to be useless variables here

Reply:
no, they have to be declared this way or else they cannot be passed to the executor service